### ISSUE TITLE
DevIno cache doesn't work when you're using a modifier on the commit

### ISSUE THREAD

**Issue by alexlarsson** [ostreedev/ostree#1165]
The way the dev-ino cache works is that if the source file matches the dev+ino, then we immediately use that file with that object as-is. However, that might not be correct if a OstreeRepoCommitModifier is in effect. The following demo demonstrates how this goes wrong:

```
# Create bare repo with a commit
mkdir to-commit
echo CONTENT > to-commit/file
ostree init --repo=repo --mode=bare
ostree commit --repo=repo --branch=master to-commit

# Check out, ensure we have hardlinks
ostree checkout --repo=repo -H master checked-out
# Check in again, but now with modified owner
sudo ostree commit --repo=repo --owner-uid=42 --branch=new-owner checked-out

# Look, these are different
ostree ls -RC --repo=repo master
ostree ls -RC --repo=repo new-owner

# Now try with --link-checkout-speedup
sudo ostree commit --repo=repo --owner-uid=42 --link-checkout-speedup --branch=new-owner-speedup checked-out

# This shows the wrong result, the owner of /file is not 42!
ostree ls -RC --repo=repo new-owner-speedup
```

I think we need to run the modifier on the actual file metadata, and only use the hardlinked file in case none of the metadata changes.

**Comment by cgwalters** [comment]
(Why `sudo ostree commit`?)

Is this a regression?  Or something new you're trying to do?

Definitely a bug anyways.

**Comment by jlebon** [comment]
Will take a look at this.

**Comment by cgwalters** [comment]
I want to understand the use case here.  Is there one or is this theoretical?

Note: the command line is mostly just a demo, and `--link-checkout-speedup` in particular should be viewed as deprecated in favor of `OstreeRepoDevInoCache` (not accessible from cmdline since it requires state).  I'm *presuming* this is just a cmdline version of a flatpak issue though - is that correct?

Under what scenarios do you want to globally change the UID?

**Comment by alexlarsson** [comment]
I do not have a particular usecase, I just noticed this when looking for ways to optimize flatpak-builder. 

In particular, it has a ostree repo for the build cache and a hardlinked checkout of that, and then we want to commit (a subset of) this directory to a different repo. I was hoping we could use something like a dev-ino cache to do map the files being commited to the cache repo and then avoiding to do the checksum of every file. That approach failed though, because the cache repo has different metadata than the final repo, so the cache checksums are wrong anyway. (The cache repo has files with your uid, but the final repo has canonical permissions.)

But, that is really unrelated to this issue, which exists in the current code irrespective of that. I agree that the demo is a bit random by just setting a uid. But this affects *any* metadata changes, such as --canonical-permissions, which is more realistic. 

Maybe we should just live with this and leave --link-checkout-speedup as an advanced feature that you have to be aware of the details to use.

**Comment by cgwalters** [comment]
Why not make the cache repo `bare-user-only`?  Then the metadata is canonicalized and should match, right?

**Comment by jlebon** [comment]
Regardless of `--link-checkout-speedup`, even in stateful shlib clients, there's definitely a use case for combining `OstreeRepoDevInoCache` with fileinfo/xattr modifications on a small subset of the files, right? If we don't want to be fancy, at least we should just completely disable the cache if any callbacks are provided.

**Comment by cgwalters** [comment]
Yeah, in fact https://github.com/projectatomic/rpm-ostree/issues/1011

**Comment by jlebon** [comment]
And that's what #1170 addresses, right?
